### PR TITLE
[preview][aptos-node] support remote config loading

### PR DESCRIPTION
**Pull request by ibalajiarun**
### Description

For previewnet, we want the nodes to download configs automatically to speed up experimentation.

It works as follows.
- Each config override is versioned.
- Node telemetry polls the telemetry service periodically for any config overrides. 
- If it can find a override config whose version is larger than the one available locally, it will verify that the override are sane and write them to a file in `<data_dir>/remote_configs` and restart the node.

On restart:
- The node checks the `<data_dir>/remote_configs` dir for the latest versioned config override file and applies it to its initial config. If the merge fails, it falls back to initial config.

Telemetry service:
- The override configs for the nodes are maintained outside of the Telemetry Service (e.g. in a Github repo or Gist) in a file. This file has a version field, the peers ids and the override config. The peer ids allow for gradual rollout of configs (so we don't cause a wildfire).
- Telemetry service periodically polls the remote config URL and downloads any updated configs and serves aptos nodes accordingly.

### CODE REVIEW

**Comment by ibalajiarun** [comment]
Current dependencies on/for this PR:
* **#12238** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/12238?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/aptos-labs/aptos-core/12238?utm_source=stack-comment).

File: aptos-node/src/lib.rs [hunk 1]
```diff
@@ -181,6 +182,70 @@ impl AptosNodeArgs {
...
+                return None;
+            };
+            let file_name = entry.file_name().into_string().ok()?;
+            if file_name.starts_with("remote_config_v") {
+                return Some(file_name);
+            }
+            None
+        })
+        .collect::<Vec<_>>();
+    entries.sort();
```
Comment (bchocho) [inline aptos-node/src/lib.rs:204]: Doesn't seem necessary since we go through all entries.

If we were to rely on this, since this is an alphabetical sort, we'll have to pad 0's. e.g., start with v000 to support 1000 versions

File: consensus/src/dag/dag_state_sync.rs [hunk 2]
```diff
@@ -108,16 +108,16 @@ impl StateSyncTrigger {
...
-                .unwrap_or_default()
-                >= ledger_info.commit_info().round()
+            .get_highest_committed_anchor_round();
+        let local_highest_ordered_round = self
+            .dag_store
+            .read()
+            .highest_ordered_anchor_round()
+            .unwrap_or_default();
+        if local_highest_committed_round < ledger_info.commit_info().round()
+        // && local_highest_ordered_round < ledger_info.commit_info().round()
```
Comment (bchocho) [inline consensus/src/dag/dag_state_sync.rs:120]: These are unrelated changes. Should they be split out to another PR?

File: aptos-node/src/lib.rs [hunk 3]
```diff
@@ -181,6 +182,70 @@ impl AptosNodeArgs {
...
+    }
+
+    if max_config.is_empty() {
+        warn!("no latest version of remote config found");
+        return Ok(None);
+    }
+
+    let latest_remote_config_path = config_dir.join(max_config);
+    let remote_config_file = fs::File::open(latest_remote_config_path)?;
+    let remote_config = serde_yaml::from_reader(remote_config_file)?;
```
Comment (bchocho) [inline aptos-node/src/lib.rs:235]: Can we just be super verbose and print out the whole remote config applied here and its version? Might help sanity check for us and operators if configs seem out of sync

File: crates/aptos-telemetry/src/service.rs [hunk 4]
```diff
@@ -221,6 +235,60 @@ fn try_spawn_log_env_poll_task(sender: TelemetrySender) {
...
+            }
+        });
+    }
+}
+
+fn process_remote_config(
+    remote_config: RemoteNodeConfig,
+    initial_node_config: &NodeConfig,
+) -> anyhow::Result<()> {
+    info!("Received new config change: {:?}", remote_config);
```
Comment (bchocho) [inline crates/aptos-telemetry/src/service.rs:262]: At this point we don't know it's "new"?

File: crates/aptos-telemetry/src/service.rs [hunk 5]
```diff
@@ -221,6 +235,60 @@ fn try_spawn_log_env_poll_task(sender: TelemetrySender) {
...
 
+fn try_spawn_remote_config_poll_task(initial_node_config: NodeConfig, sender: TelemetrySender) {
+    if enable_remote_config_polling() {
+        tokio::spawn(async move {
+            let mut interval = time::interval(Duration::from_secs(REMOTE_CONFIG_POLL_FREQ_SECS));
+            loop {
+                interval.tick().await;
+                if let Some(remote_config) = sender.get_remote_config().await {
+                    if let Err(err) = process_remote_config(remote_config, &initial_node_config) {
+                        warn!("error processing remote config: {}", err);
```
Comment (bchocho) [inline crates/aptos-telemetry/src/service.rs:246]: This doesn't seem like an error if the file exists, it's expected?

File: crates/aptos-telemetry-service/src/types/telemetry.rs [hunk 6]
```diff
@@ -29,3 +31,23 @@ pub(crate) struct BigQueryRow {
...
+    pub node_type: NodeType,
+    pub chain_id: ChainId,
+    pub peer_ids: Vec<PeerId>,
+    pub node_config: serde_yaml::Value,
+}
+
+#[derive(Debug, Deserialize, Clone)]
+pub struct RemoteNodeConfigsFile {
+    pub version: u64,
+    pub node_configs: Vec<RemoteNodeConfigEntry>,
```
Comment (bchocho) [inline crates/aptos-telemetry-service/src/types/telemetry.rs:46]: Shouldn't the node configs file contain a list of remote config updates? E.g., if rolling out to nodes in three stages, won't you want to append in case some validators didn't get the update yet, or wiped their disks? Maybe it's overkill.
